core/stat/base/leap_array.go

```diff
@@ -0,0 +1,203 @@
+package base
+
+import (
+	"fmt"
+	"github.com/pkg/errors"
+	"github.com/sentinel-group/sentinel-golang/util"
+	"reflect"
+	"runtime"
+	"sync/atomic"
+	"unsafe"
+)
+
+const (
+	PtrSize = int(8)
+)
+
+// windowWrap represent a slot to record metrics
+// In order to reduce the usage of memory, windowWrap don't hold length of windowWrap
+// The length of windowWrap could be seen in leapArray.
+// The scope of time is [startTime, startTime+windowLength)
+// The size of windowWrap is 24(8+16) bytes
+type windowWrap struct {
+	// Start time of this windowWrap
+	windowStart uint64
+	// Value is the actual data structure to record the metrics.
+	// Such as metricBucket
+	value interface{}
+}
+
+func (ww *windowWrap) resetTo(startTime uint64) {
+	ww.windowStart = startTime
+}
+
+func (ww *windowWrap) isTimeInWindow(timeMillis uint64, windowLengthInMs uint32) bool {
+	return ww.windowStart <= timeMillis && timeMillis < ww.windowStart+uint64(windowLengthInMs)
+}
+
+func calculateStartTime(now uint64, windowLengthInMs uint32) uint64 {
+	return now - (now % uint64(windowLengthInMs))
+}
+
+// atomic windowWrap array to resolve race condition
+// atomicWindowWrapArray can not append or delete element after initializing
+type atomicWindowWrapArray struct {
+	// The base address for real data array
+	base uintptr
```

Comment:
Caching `uintptr` is dangerous and is highly not recommended according to the document of `unsafe` package:

> Converting a Pointer to a uintptr produces the memory address of the value pointed at, as an integer. The usual use for such a uintptr is to print it.
> Conversion of a uintptr back to Pointer is not valid in general.
> A uintptr is an integer, not a reference. Converting a Pointer to a uintptr creates an integer value with no pointer semantics. Even if a uintptr holds the address of some object, the garbage collector will not update that uintptr's value if the object moves, nor will that uintptr keep the object from being reclaimed.

We'd better use the `unsafe.Pointer` instead.

References:

- https://golang.org/pkg/unsafe/
- https://go101.org/article/unsafe.html

Reply:
That's cool,

> In general, reflect.SliceHeader and reflect.StringHeader should be used only as *reflect.SliceHeader and *reflect.StringHeader pointing at actual slices or strings, never as plain structs. A program should not declare or allocate variables of these struct types.

I also replace the reflect.StringHeader with safe version. Please see below：
```go
type SliceHeader struct {
	Data unsafe.Pointer
	Len  int
	Cap  int
}
```